docs/plugin-authoring-evaluator.md

```diff
@@ -0,0 +1,380 @@
+# Writing a Genkit Evaluator Plugin
+
+Firebase Genkit can be extended to support custom use cases that leverage an LLM as a judge or a purely programatic evaluation of test case output.
+
+## Evaluator definition
+
+Evaluators are functions that assess the content given to and generated by an LLM. There are two main approaches to automated evaluation (testing) - hueristic assesment and LLM-based assesment. In a hueristic approach, define a deterministic function much like those of traditional software development. In an LLM-based assesment, the content is fed back to an LLM and the LLM is asked to score the output according to criteria set in a prompt.
+
+### LLM based Evaluators
+
+An LLM-based evaluator leverages an LLM to evaluate the input, context, or output of your genAI feature.
+
+LLM-based evaluators in Genkit are made up of 3 components:
+
+- A prompt
+- A scoring function
+- An evaluator action
+
+#### Define the prompt
+
+For this example, the prompt is going to ask the LLM to judge how delicious the output is. First, provide context to the LLM, then describe what you want it to do, and finally, give it a few examples to base its response on.
+
+```ts
+const QUESTION_GEN_PROMPT = `You are a food critic with a wide range in taste. Given the output, decide if it sounds delicious and provide your reasoning. Use only "yes" (if delicous), "no" (if not delicious), "maybe" (if you can't decide) as the verdict.
+
+Here are a few examples:
+
+Output:
+Chicken parm sandwich
+Response:
+{ "reason": "This is a classic sandwich enjoyed by many - totally delicious", "verdict":"yes"}
+
+Output:
+Boston logan international airport tarmac
+Response:
+{ "reason": "This is not edible and definitely not delicious.", "verdict":"no"}
+
+Output:
+A juicy peice of gossip
+Response:
+{ "reason": "Gossip is sometimes metaphorically referred to as tasty.", "verdict":"maybe"}
+
+Here is a new submission to assess:
+
+Output:
+{{output}}
+Response:
+`;
+```
+
+#### Define the scoring function
+
+Now, define the function that will take an example which includes `output` as is required by the prompt and score the result. Genkit test cases include `input` and `output` as required fields, with optional fields for `context`. It is the responsibility of the evaluator to validate that all fields required for evaluation are present.
+
+This example leverages `handlebars` to hydrate the prompt and `zod` to format and validate the response.
```

Comment:
We should probably update the evaluators to use dotprompt instead.

Reply:
Totally agree!